### PR TITLE
CrashDump 2.0

### PR DESCRIPTION
**Pull request by bugobliterator**
This PR include changed to CrashDebug and mri to allow for doing ChibiOS per thread stack trace.

Currently there is a bug which is setting first program counter to point towards fault address. But the stack trace below that is correct for each thread.

For example:
```
  Id   Target Id                                         Frame 
* 1    Thread 805426916 (Name: ArduPlane State: CURRENT) 0xe000ed38 in ?? ()
  2    Thread 805427328 (Name: idle State: READY)        0x000817fa in ?? ()
  3    Thread 805559216 (Name: UART_RX State: READY)     0x000817fa in ?? ()
  4    Thread 805554960 (Name: OTG1 State: READY)        0x000817fa in ?? ()
  5    Thread 805411472 (Name: monitor State: SLEEPING)  0x000817fa in ?? ()
  6    Thread 805417136 (Name: timer State: SLEEPING)    0x000817fa in ?? ()
  7    Thread 805413792 (Name: rcout State: WTOREVT)     0x000817fa in ?? ()
  8    Thread 805412888 (Name: rcin State: READY)        0x000817fa in ?? ()
  9    Thread 805410568 (Name: io State: READY)          0x000817fa in ?? ()
  10   Thread 805415208 (Name: storage State: READY)     0x000817fa in ?? ()
  11   Thread 805553480 (Name: UART2 State: READY)       0x000817fa in ?? ()
  12   Thread 805548448 (Name: UART3 State: READY)       0x000817fa in ?? ()
  13   Thread 805545664 (Name: UART4 State: READY)       0x000817fa in ?? ()
  14   Thread 805542880 (Name: UART8 State: READY)       0x000817fa in ?? ()
  15   Thread 805540096 (Name: UART7 State: READY)       0x000817fa in ?? ()
  16   Thread 805536288 (Name: OTG2 State: READY)        0x000817fa in ?? ()
  17   Thread 805533040 (Name: UART6 State: WTOREVT)     0x000817fa in ?? ()
  18   Thread 805492568 (Name: IOMCU State: WTOREVT)     0x000817fa in ?? ()
  19   Thread 805516664 (Name: uavcan_0 State: READY)    0x000817fa in ?? ()
  20   Thread 805518496 (Name: SPI4 State: SLEEPING)     0x000817fa in ?? ()
  21   Thread 805520160 (Name: SPI1 State: SUSPENDED)    0x000817fa in ?? ()
  22   Thread 805523736 (Name: log_io State: READY)      0x000817fa in ?? ()
--Type <RET> for more, q to quit, c to continue without paging--
  23   Thread 805449912 (Name: FTP State: READY)         0x000817fa in ?? ()
```
```
(gdb) thread 1
[Switching to thread 1 (Thread 805426916)]
#0  0xe000ed38 in ?? ()
(gdb) bt
#0  0xe000ed38 in ?? ()
#1  0x0809e270 in GCS_MAVLINK::handle_preflight_reboot (this=0x30037db8, packet=...) at ../../libraries/GCS_MAVLink/GCS_Common.cpp:2783
#2  0x0809f036 in GCS_MAVLINK::handle_command_long_packet (this=this@entry=0x30037db8, packet=...) at ../../libraries/GCS_MAVLink/GCS_Common.cpp:4227
#3  0x080253e6 in GCS_MAVLINK_Plane::handle_command_long_packet (this=0x30037db8, packet=...) at ../../ArduPlane/GCS_Mavlink.cpp:1082
#4  0x0809c966 in GCS_MAVLINK::handle_command_long (this=0x30037db8, msg=...) at ../../libraries/GCS_MAVLink/GCS_Common.cpp:4422
#5  0x080a113e in GCS_MAVLINK::handle_common_message (this=this@entry=0x30037db8, msg=...) at ../../libraries/GCS_MAVLink/GCS_Common.cpp:3483
#6  0x08025858 in GCS_MAVLINK_Plane::handleMessage (this=0x30037db8, msg=...) at ../../ArduPlane/GCS_Mavlink.cpp:1294
#7  0x0809d136 in GCS_MAVLINK::packetReceived (this=this@entry=0x30037db8, status=..., msg=...) at ../../libraries/GCS_MAVLink/GCS_Common.cpp:1430
#8  0x08024f6e in GCS_MAVLINK_Plane::packetReceived (this=0x30037db8, status=..., msg=...) at ../../ArduPlane/GCS_Mavlink.cpp:641
#9  0x0809be3e in GCS_MAVLINK::update_receive (this=0x30037db8, max_time_us=max_time_us@entry=1000) at ../../libraries/GCS_MAVLink/GCS_Common.cpp:1481
#10 0x0809edb2 in GCS::update_receive (this=0x30007010 <plane+16384>) at ../../libraries/GCS_MAVLink/GCS_Common.cpp:2136
#11 0x080222ee in Functor<void>::method_wrapper<GCS, &GCS::update_receive> (obj=<optimized out>) at ../../libraries/AP_HAL/utility/functor.h:88
#12 0x0808f5a4 in Functor<void>::operator() (this=0x818b27c <Plane::scheduler_tasks+360>) at ../../libraries/AP_HAL/utility/functor.h:52
#13 AP_Scheduler::run (this=this@entry=0x30003238 <plane+552>, time_available=19635, time_available@entry=19999) at ../../libraries/AP_Scheduler/AP_Scheduler.cpp:198
#14 0x0808fac6 in AP_Scheduler::loop (this=this@entry=0x30003238 <plane+552>) at ../../libraries/AP_Scheduler/AP_Scheduler.cpp:323
#15 0x08091906 in AP_Vehicle::loop (this=0x30003010 <plane>) at ../../libraries/AP_Vehicle/AP_Vehicle.cpp:182
#16 0x080ee18c in main_loop () at ../../libraries/AP_HAL_ChibiOS/HAL_ChibiOS_Class.cpp:266
#17 0x080ee1cc in HAL_ChibiOS::run (this=<optimized out>, argc=<optimized out>, argv=<optimized out>, callbacks=0x30014340 <AP_HAL::get_HAL()::hal_chibios>)
    at ../../libraries/AP_HAL_ChibiOS/HAL_ChibiOS_Class.cpp:315
#18 0x08023008 in main (argc=<optimized out>, argv=<optimized out>) at ../../ArduPlane/ArduPlane.cpp:764
```

```
Thread 4 (Thread 805554960 (Name: OTG1 State: READY)):
#0  0x000817fa in ?? ()
#1  0x0809e270 in GCS_MAVLINK::handle_preflight_reboot (this=0x0, packet=...) at ../../libraries/GCS_MAVLink/GCS_Common.cpp:2783
#2  0x0817fa82 in chSchGoSleepTimeoutS (newstate=newstate@entry=10 '\n', timeout=<optimized out>) at ../../modules/ChibiOS/os/rt/src/chschd.c:266
#3  0x0817f790 in chEvtWaitAnyTimeout (events=10, events@entry=4108, timeout=timeout@entry=1000) at ../../modules/ChibiOS/os/rt/src/chevents.c:550
#4  0x081677be in ChibiOS::UARTDriver::uart_thread (this=0x3000b5c8 <uartADriver>) at ../../modules/ChibiOS/os/rt/include/chtime.h:356
#5  0x081677f6 in ChibiOS::UARTDriver::uart_thread_trampoline (p=<optimized out>) at ../../libraries/AP_HAL_ChibiOS/UARTDriver.cpp:205
#6  0x08020f2a in _port_thread_start ()
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
```

`GCS_MAVLINK::handle_preflight_reboot` is somehow finding its way into the stack trace for thread 4 as well. Everything below that is sane though. Its mildly annoying but will need find out why this is.

Also including move to using all of remaining space in flash to log as much as we can starting with main stack, per thread stack, then Statically allocated memory and finally heap.

One more step towards completing https://github.com/ArduPilot/ardupilot/issues/18674

### CODE REVIEW

File: libraries/AP_Filesystem/AP_Filesystem_Sys.cpp [hunk 1]
```diff
@@ -166,8 +170,12 @@ int32_t AP_Filesystem_Sys::read(int fd, void *buf, uint32_t count)
         return -1;
     }
     struct rfile &r = file[fd];
-    count = MIN(count, r.str->get_length() - r.file_ofs);
-    memcpy(buf, &r.str->get_string()[r.file_ofs], count);
+    if (r.is_crash_dump) {
```
Comment (tridge) [inline libraries/AP_Filesystem/AP_Filesystem_Sys.cpp:173]: I suggest we do this:
 - add set_data_const(const uint8_t *data, uint32_t size) to ExpandingString
 - that would mark as "don't free" in ExpandingString destructor
 - add a get_storage_ptr() fn to hal.storage for getting storage ptr
it means no more special cases

**Review by tridge — CHANGES_REQUESTED** [review]
fantastic to get the threads!
just a suggestion on putting the "don't free" inside ExpandingString

File: libraries/AP_Filesystem/AP_Filesystem_Sys.cpp (hunk 1 above)
Comment (bugobliterator) [reply libraries/AP_Filesystem/AP_Filesystem_Sys.cpp:173]: ok, will do.

File: .gitmodules [hunk 2]
```diff
@@ -33,4 +33,7 @@
 	url = https://github.com/DroneCAN/DSDL.git
 [submodule "modules/CrashDebug"]
 	path = modules/CrashDebug
-	url = https://github.com/adamgreen/CrashDebug
+	url = https://github.com/ardupilot/CrashDebug
+[submodule "modules/esp_idf"]
```
Comment (davidbuzz) [inline .gitmodules:37]: the esp_idf submodule shouldn't be here, we removed it.

**Review by tridge — APPROVED** [review]

**Pull request merged**